File: src/leetcode/p0000_0099/p0070_climbing_stairs/ClimbingStairs.java
/*
 * Copyright (C) 2021  Weiyang(Will) Xu
 * 
 * This file is part of algorithm.
 * Algorithm is free software: you can redistribute it and/or modify
 * it under the term of the GNU General Public License version 3 or
 * any later version, as specified in the readme.md file.
 */

package org.willxu.algorithm.service.integer;

public interface ClimbingStairs {

    /**
     * You are climbing a staircase. It takes n steps to reach the top.
     *
     * Each time you can either climb 1 or 2 steps. In how many
     * distinct ways can you climb to the top?
     *
     * @param n 1 <= n <= 45
     * @return number of distinct ways to climb to the top
     */
    int climbStairs(int n);
}

File: src/leetcode/p0000_0099/p0070_climbing_stairs/climbing_stairs.py
from abc import ABC, abstractmethod

class ClimbingStairs(ABC):
    """
    You are climbing a staircase. It takes n steps to reach the top.
    Each time you can either climb 1 or 2 steps. In how many distinct ways can you climb to the top?
    """
    @abstractmethod
    def climb_stairs(self, n: int) -> int:
        """
        Args:
            n:
                1 <= n <= 45
        """
        pass

class ClimbingStairsDp(ClimbingStairs):
    def climb_stairs(self, n: int) -> int:
        if n <= 2:
            return n
        dp = [0] * (n + 1)
        dp[1], dp[2] = 1, 2
        for i in range(3, n + 1):
            dp[i] = dp[i - 1] + dp[i - 2]
        return dp[n]
